woocommerce/woocommerce-ios
# REST API: New credential type for application password

Part of #9186 


## Description

This PR adds support for logging in to the app after authorizing application password through wp-admin. Changes include:
- Networking:
  - Added a new case for `Credentials` with wporg username and application password.
  - Added `OneTimeApplicationPasswordUseCase` to handle application password after authorizing through wp-admin. The use case saves the password into storage and will not handle re-generation. Deletion is handled by fetching the UUID of the matching application password before removal. All network requests are handled with `URLSession` for simplicity.
  - Updated `RequestAuthenticator` and `RequestConverter` to help `AlamofireNetwork` authenticate requests when the user is authenticated with application password.
- UI:
  - Fixed issue logging out users when application password generation fails by listening to the failure notification from `AppDelegate`. 
  - Updated the navigation type for application password authorization web view to push it to a navigation controller instead of presenting modally.
  - Updated `AuthenticationManager` to authenticate the user with application password after authorization and trigger post-login checks.
  - Updated the store widget to make use of the new credential type.

## Testing instructions

Prerequisite: Make sure that you have access to a self-hosted site with Woo but not Jetpack.
- Install and configure [login captcha](https://wordpress.org/plugins/login-recaptcha/) on your test site.
- Build the app and attempt log in to your site with the correct credentials.
- An alert should be displayed saying the login fails. Tap the CTA to log in through wp-admin.
- Log in and authorize application password in the presented web view. Notice that the app then logs in successfully and navigates to the home screen.
- Notice that all REST requests should still succeed.
- Switch to the `StoreWidgetsExtension` target, build, and run. Notice that the widget can still load store stats properly.
- Switch to the app, navigate to Settings to log out.
- Check your user in wp-admin, notice that the authorized application password has been deleted successfully.

Testing auto-logout:
- Repeat the steps in the previous test to log in to your test site through the web view.
- Go to wp-admin and revoke the password that you just authorized.
- On the app, navigate to some screen that involves making API requests. Notice that you are logged out immediately.


## Screenshots


https://user-images.githubusercontent.com/5533851/226531266-92096271-b5b3-460f-8d6d-5c6da78be74e.mp4



---
